dll/win32/shell32/wine/shellord.c

```diff
@@ -403,6 +432,58 @@ int ShellMessageBoxA(
 	UINT uType,
 	...)
 {
+#ifdef __REACTOS__
+    CHAR *szText = NULL, szTitle[100];
+    LPCSTR pszText, pszTitle = szTitle;
+    LPSTR pszTemp;
+    __ms_va_list args;
+    int ret;
+
+    __ms_va_start(args, uType);
+
+    TRACE("(%p,%p,%p,%p,%08x)\n", hInstance, hWnd, lpText, lpCaption, uType);
+
+    if (IS_INTRESOURCE(lpCaption))
+        LoadStringA(hInstance, LOWORD(lpCaption), szTitle, ARRAY_SIZE(szTitle));
+    else
+        pszTitle = lpCaption;
+
+    if (IS_INTRESOURCE(lpText))
+    {
+        /* Retrieve the maximum possible length of the corresponding
+         * Unicode string (not counting any possible NULL-terminator). */
+        const WCHAR *ptr;
+        UINT len = LoadStringW(hInstance, LOWORD(lpText), (LPWSTR)&ptr, 0);
```

Comment:
It's better to use the A here as well.

Reply:
It's actually not possible to call the LoadStringA with this set of parameters (both on Windows and ReactOS) in order to just retrieve the size of the string, without using a buffer to also retrieve the string contents (not the pointer).

Reply:
yes, so give it a length of 1 or so, and a dummy buffer so it can return the actual length?
The length in WCHAR is most likely not the same as the length in Ansi..

Reply:
or alternatively, convert the string yourself (and calculate resulting length),
but then you'd have to do this the same way as `LoadStringA` is doing.

Reply:
Documentation says:

> If the function succeeds, the return value is one of the following:
>
> - **The number of characters copied into the buffer (if cchBufferMax is non-zero), not including the terminating null character.**
>
> - The number of characters in the string resource that lpBuffer points to (if cchBufferMax is zero). The string resource is not guaranteed to be null-terminated in the module's resource table, and you can use this value to determine where the string resource ends.
>
> - Zero if the string resource does not exist.
>
> To get extended error information, call GetLastError.

Our implementation of LoadStringA (which is similar to the one of Windows) is:
https://git.reactos.org/?p=reactos.git;a=blob;f=win32ss/user/user32/misc/resources.c;hb=f3bc1b396aa5471cc4eb820ce4586c0ca5b6d1a8#l77

So first of all it expects a non-zero cchBufferMax (the buflen parameter; note that it doesn't check for the buffer pointer to be NULL); then, when the string is found, it translates it in ANSI into the passed buffer, taking into account its size from the buflen parameter, and returns in retval the size of the copied and converted string part.

Reply:
Yes, that's the discussion we had in my previous PR. Looks like our implementation of LoadStringA is not as per MSDN

Reply:
Our implementation has to be on par with what Windows actually does, and not what MSDN may say if it's wrong.

Reply:
(For record, there are winetests that check this behaviour:
https://git.reactos.org/?p=reactos.git;a=blob;f=modules/rostests/winetests/user32/resource.c;hb=5077e400d5a92d15516c0dd1c7fa4cdb9399dd12#l125
)

Reply:
I calculate the more correct necessary character number using the WideCharToMultiByte() call below.